File: frontend/app/api/holos.ts
import { fetchWithAuth } from "./auth";

const RAW_API_BASE = (import.meta as any).env?.VITE_API_BASE_URL || "http://localhost:5001";
const API_BASE_URL = RAW_API_BASE.replace(/\/$/, "");

export interface HoloConfig {
  holo_id: string;
  user_id: string;
  questions: string[];
}

export interface HoloDaily {
  holo_daily_id: string;
  holo_id: string;
  entry_date: string;
  score: number;
  answers: Record<string, boolean>;
}

export interface HoloDailyCreate {
  entry_date: string;
  score: number;
  answers: Record<string, boolean>;
}

export interface AvgScoreResponse {
  avg_score: number | null;
}

// Get holo config for current user
export async function getHoloConfig(): Promise<HoloConfig> {
  console.log('🌐 getHoloConfig: Starting...');
  console.log('🌐 getHoloConfig: URL:', `${API_BASE_URL}/holos/holo`);
  
  const response = await fetchWithAuth(`${API_BASE_URL}/holos/holo`);
  console.log('🌐 getHoloConfig: Response status:', response.status);

  if (!response.ok) {
    if (response.status === 404) {
      console.log('🌐 getHoloConfig: 404 - No config found');
      throw new Error('No holo configuration found');
    }
    const errorText = await response.text();
    console.error('🌐 getHoloConfig: Error response:', errorText);
    throw new Error(`Failed to fetch holo config: ${response.status} ${response.statusText} - ${errorText}`);
  }

  const result = await response.json();
  console.log('🌐 getHoloConfig: Success:', result);
  return result;
}

// Create holo config for current user
export async function createHoloConfig(questions: string[]): Promise<HoloConfig> {
  const response = await fetchWithAuth(`${API_BASE_URL}/holos/holo`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ questions }),
  });

  if (!response.ok) {
    throw new Error(`Failed to create holo config: ${response.statusText}`);
  }

  return response.json();
}

// Get holo daily by date
export async function getHoloDaily(entryDate: string): Promise<HoloDaily | null> {
  const response = await fetchWithAuth(`${API_BASE_URL}/holos/daily?entry_date=${entryDate}`);

  if (response.status === 404) {
    return null;
  }

  if (!response.ok) {
    throw new Error(`Failed to fetch holo daily: ${response.statusText}`);
  }

  return response.json();
}

// Get latest holo daily
export async function getLatestHoloDaily(): Promise<HoloDaily | null> {
  const response = await fetchWithAuth(`${API_BASE_URL}/holos/daily/latest`);

  if (response.status === 404) {
    return null;
  }

  if (!response.ok) {
    throw new Error(`Failed to fetch latest holo daily: ${response.statusText}`);
  }

  return response.json();
}

// Create holo daily
export async function createHoloDaily(holoDaily: HoloDailyCreate): Promise<HoloDaily> {
  const response = await fetchWithAuth(`${API_BASE_URL}/holos/daily`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(holoDaily),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Failed to create holo daily: ${response.status} ${response.statusText} - ${errorText}`);
  }

  return response.json();
}

// Get average score
export async function getAvgScore(): Promise<AvgScoreResponse> {
  const response = await fetchWithAuth(`${API_BASE_URL}/holos/avg-score`);

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Failed to fetch average score: ${response.status} ${response.statusText} - ${errorText}`);
  }

  return response.json();
}

File: backend/src/api/routes/holos.py
from fastapi import APIRouter, Depends, HTTPException
from typing import List
from datetime import date
from sqlalchemy.orm import Session
from sqlalchemy.exc import SQLAlchemyError, IntegrityError
from pydantic import ValidationError

from src.models.holos import Holo, HoloCreate, HoloUpdate, HoloDaily, HoloDailyCreate
from src.db.holos import (
    get_holo_config,
    update_holo_config,
    create_holo_config,
    get_holo_daily_by_date,
    get_latest_holo_daily,
    create_holo_daily,
    get_avg_score,
)
from src.db.session import get_db
from src.api.routes.auth import get_current_user

router = APIRouter(prefix="/holos", tags=["holos"])


@router.get("/holo", response_model=Holo)
def get_holo_config_route(
    db: Session = Depends(get_db), user=Depends(get_current_user)
):
    """Get the holo questions config for a user"""
    try:
        result = get_holo_config(user["uid"], db)
        if result is None:
            raise HTTPException(status_code=404, detail="Holo configuration not found")
        return result
    except HTTPException:
        raise  # Re-raise HTTPException as-is
    except SQLAlchemyError as e:
        raise HTTPException(
            status_code=500,
            detail=f"Database error while fetching holo config: {str(e)}",
        )
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Unexpected error while fetching holo config: {str(e)}",
        )


@router.put("/holo", response_model=Holo)
def update_holo_config_route(
    holo: HoloUpdate, db: Session = Depends(get_db), user=Depends(get_current_user)
):
    """Update the holo questions config for the current user"""
    try:
        result = update_holo_config(user["uid"], holo, db)
        if result is None:
            raise HTTPException(status_code=404, detail="Holo configuration not found")
        return result
    except HTTPException:
        raise  # Re-raise HTTPException as-is
    except ValidationError as e:
        raise HTTPException(status_code=422, detail=f"Validation error: {str(e)}")
    except SQLAlchemyError as e:
        raise HTTPException(
            status_code=500,
            detail=f"Database error while updating holo config: {str(e)}",
        )
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Unexpected error while updating holo config: {str(e)}",
        )


@router.post("/holo", response_model=Holo)
def create_holo_config_route(
    holo: HoloCreate, db: Session = Depends(get_db), user=Depends(get_current_user)
):
    """Create a new holo questions config for a user - TODO triggered on signup"""
    try:
        return create_holo_config(user["uid"], holo, db)
    except ValidationError as e:
        raise HTTPException(status_code=422, detail=f"Validation error: {str(e)}")
    except IntegrityError as e:
        raise HTTPException(status_code=400, detail=f"Data integrity error: {str(e)}")
    except SQLAlchemyError as e:
        raise HTTPException(
            status_code=500,
            detail=f"Database error while creating holo config: {str(e)}",
        )
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Unexpected error while creating holo config: {str(e)}",
        )


@router.get(
    "/daily",
    response_model=HoloDaily,
    description="Get the holo daily by date for a user",
)
def get_holo_daily_route(
    entry_date: date, db: Session = Depends(get_db), user=Depends(get_current_user)
):
    """Get the holo daily for a user"""
    try:
        holo = get_holo_config(user["uid"], db)
        if not holo:
            raise HTTPException(404, "No holo config found")
        result = get_holo_daily_by_date(holo.holo_id, entry_date, db)
        if not result:
            raise HTTPException(404, "Holo daily not found")
        return result
    except HTTPException:
        raise  # Re-raise HTTPException as-is
    except ValidationError as e:
        raise HTTPException(status_code=422, detail=f"Validation error: {str(e)}")
    except IntegrityError as e:
        raise HTTPException(status_code=400, detail=f"Data integrity error: {str(e)}")
    except SQLAlchemyError as e:
        raise HTTPException(
            status_code=500,
            detail=f"Database error while fetching holo daily: {str(e)}",
        )
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Unexpected error while fetching holo daily: {str(e)}",
        )


@router.get("/daily/latest", response_model=HoloDaily)
def get_latest_holo_daily_route(
    db: Session = Depends(get_db), user=Depends(get_current_user)
):
    """Get the latest holo daily for a user"""
    try:
        holo = get_holo_config(user["uid"], db)
        if not holo:
            raise HTTPException(404, "No holo config found")
        latest_holo = get_latest_holo_daily(holo.holo_id, db)
        if not latest_holo:
            raise HTTPException(404, "No holo daily found")
        return latest_holo
    except HTTPException:
        raise  # Re-raise HTTPException as-is
    except ValidationError as e:
        raise HTTPException(status_code=422, detail=f"Validation error: {str(e)}")
    except IntegrityError as e:
        raise HTTPException(status_code=400, detail=f"Data integrity error: {str(e)}")
    except SQLAlchemyError as e:
        raise HTTPException(
            status_code=500,
            detail=f"Database error while fetching latest holo daily: {str(e)}",
        )
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Unexpected error while fetching latest holo daily: {str(e)}",
        )


@router.post("/daily", response_model=HoloDaily)
def create_holo_daily_route(
    holo_daily: HoloDailyCreate,
    db: Session = Depends(get_db),
    user=Depends(get_current_user),
):
    """Create a new holo daily for a user"""
    try:
        holo = get_holo_config(user["uid"], db)
        if not holo:
            raise HTTPException(404, "No holo config found")
        return create_holo_daily(holo.holo_id, holo_daily, db)
    except HTTPException:
        raise  # Re-raise HTTPException as-is
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=422, detail=f"Validation error: {str(e)}")
    except IntegrityError as e:
        raise HTTPException(status_code=400, detail=f"Data integrity error: {str(e)}")
    except SQLAlchemyError as e:
        raise HTTPException(
            status_code=500,
            detail=f"Database error while creating holo daily: {str(e)}",
        )
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Unexpected error while creating holo daily: {str(e)}",
        )


@router.get("/avg-score")
def get_avg_score_route(db: Session = Depends(get_db), user=Depends(get_current_user)):
    """Get the average score from all holo dailies for a user"""
    try:
        holo = get_holo_config(user["uid"], db)
        if not holo:
            raise HTTPException(404, "No holo config found")

        avg_score = get_avg_score(holo.holo_id, db)
        return {"avg_score": avg_score}
    except HTTPException:
        raise  # Re-raise HTTPException as-is
    except SQLAlchemyError as e:
        raise HTTPException(
            status_code=500,
            detail=f"Database error while fetching average score: {str(e)}",
        )
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Unexpected error while fetching average score: {str(e)}",
        )
